graylog2-server/src/main/java/org/graylog2/plugin/Message.java

```diff
@@ -953,4 +962,40 @@ public String apply(final Message input) {
             return input.getId();
         }
     }
+
+    /**
+     * Store the specified metadata value within the message's internal metadata map.
+     *
+     * @param key   A globally unique string key identifier for the metadata value.
+     * @param value The metadata object value.
+     */
+    public void setMetadata(String key, Object value) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(key), "A non-empty key is required.");
+        Preconditions.checkNotNull(value);
+        if (metadata == null) {
+            metadata = new HashMap<>();
+        }
+    }
+
+    /**
+     * Get the metadata value for the specified key.
+     *
+     * @param key The string key for the metadata entry.
+     */
+    public Object getMetadataValue(String key) {
+        return metadata.get(key);
+    }
```

Comment:
How about annotating this with `@Nullable`? This will help during development.

What do you think about adding a `getMetadataValue(String key, Object defaultValue)` method. Might be useful to avoid null checks in consumers of the class.

Reply:
We also need a check if the metadata is `null` to avoid a NullPointerException when trying to get a value from an uninitialized map.

Reply:
This all sounds good to me.